rpc_state_reader_sn_api/src/lib.rs

```diff
@@ -0,0 +1,994 @@
+use starknet_api::hash::StarkFelt;
+use starknet_api::serde_utils::PrefixedBytesAsHex;
+use starknet_api::transaction::{InvokeTransaction, Transaction, TransactionHash};
+use starknet_api::{core::ContractAddress, hash::StarkHash, state::StorageKey};
+use std::collections::HashMap;
+use std::env;
+use std::sync::Arc;
+use thiserror::Error;
+
+/// Starknet chains supported in Infura.
+#[derive(Debug, Clone, Copy)]
+pub enum RpcChain {
+    MainNet,
+    TestNet,
+    TestNet2,
+}
+
+impl fmt::Display for RpcChain {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            RpcChain::MainNet => write!(f, "starknet-mainnet"),
+            RpcChain::TestNet => write!(f, "starknet-goerli"),
+            RpcChain::TestNet2 => write!(f, "starknet-goerli2"),
+        }
+    }
+}
+
+/// A [StateReader] that holds all the data in memory.
+///
+/// This implementation is uses HTTP requests to call the RPC endpoint,
+/// using Infura.
+/// In order to use it an Infura API key is necessary.
+pub struct RpcState {
+    /// Enum with one of the supported Infura chains/
+    chain: RpcChain,
+    /// Infura API key.
+    api_key: String,
+    /// Struct that holds information on the block where we are going to use to read the state.
+    block: BlockValue,
+}
+
+#[derive(Debug, Error)]
+enum RpcError {
+    #[error("RPC call failed with error: {0}")]
+    RpcCall(String),
+    #[error("Request failed with error: {0}")]
+    Request(String),
+    #[error("Failed to cast from: {0} to: {1} with error: {2}")]
+    Cast(String, String, String),
+}
+
+/// [`BlockValue`] is an Enum that represent which block we are going to use to retrieve information.
+#[allow(dead_code)]
+pub enum BlockValue {
+    /// String one of: ["latest", "pending"]
+    Tag(serde_json::Value),
+    /// Integer
+    Number(serde_json::Value),
+    /// String with format: 0x{felt252}
+    Hash(serde_json::Value),
+}
+
+impl BlockValue {
+    fn to_value(&self) -> serde_json::Value {
+        match self {
+            BlockValue::Tag(block_tag) => block_tag.clone(),
+            BlockValue::Number(block_number) => json!({ "block_number": block_number }),
+            BlockValue::Hash(block_hash) => json!({ "block_hash": block_hash }),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct RpcResponseContractClass {
+    result: SNContractClass,
+}
+
+// We use this new struct to cast the string that contains a [`StarkFelt`] in hex to a [`StarkFelt`]
+struct StarkFeltHex;
+
+impl<'de> DeserializeAs<'de, StarkFelt> for StarkFeltHex {
+    fn deserialize_as<D>(deserializer: D) -> Result<StarkFelt, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let bytes: PrefixedBytesAsHex<32> = PrefixedBytesAsHex::deserialize(deserializer)?;
+        Ok(bytes.try_into().unwrap())
+    }
+}
+
+#[serde_as]
+#[derive(Debug, Deserialize)]
+struct RpcResponseFelt {
+    #[serde_as(as = "StarkFeltHex")]
+    result: StarkFelt,
```

Comment:
I think this would work with [`#[serde(deserialize_with = "...")]`](https://serde.rs/field-attrs.html#deserialize_with) just fine, without requiring another 3rd party library (`serde_with`).

Reply:
This isn't even necessary, `StarkFelt` already implements deserialization via `PrefixedBytesAsHex`. Thanks!